// @maersk-global/shared-js/jest.config.js
module.exports = {
  // Indicates whether the coverage information should be collected while executing the test
  collectCoverage: true,

  // An array of glob patterns indicating a set of files for which coverage information should be collected
  collectCoverageFrom: ["src/**/*.{js,ts}", "!src/**/index.js"],

  // The directory where Jest should output its coverage files
  coverageDirectory: "coverage",

  // A set of global variables that need to be available in all test environments
  // globals: {},

  globals: {
    "ts-jest": {
      // babelConfig: true,
      // tsConfig: "tsconfig.spec.json"
      tsconfig: {
        compilerOptions: {
          strict: false
        }
      }
    }
  },
  // An array of directory names to be searched recursively up from the requiring module's location
  moduleDirectories: ["node_modules", "src"],

  // An array of file extensions your modules use
  moduleFileExtensions: ["js", "ts", "tsx"],

  // The test environment that will be used for testing
  testEnvironment: "jest-environment-jsdom-global"
};
