// @maersk-global/shared-js/rollup.config.js
// rollup.config.js

/**
 * This is the build process that produces easier-to-use libraries
 * that mean you won't have to transpile your dependencies.
 *
 * */
import resolve from "@rollup/plugin-node-resolve";
import babel from "rollup-plugin-babel";
import typescript from "@rollup/plugin-typescript";

// get the format it's being run under
const format = process.argv[process.argv.indexOf("-f") + 1];

const plugins = [];

if (format === "iife") {
  plugins.push(
    // Note - we don't use the resolve plugin because it causes the whole of core-js to be bundled in the library
    // and webpack is unable to tree-shake it out. This would be needed if we did a browser version, but we're not going to
    resolve({
      // pass custom options to the resolve plugin so that it resolves where modules are keps
      customResolveOptions: {
        moduleDirectory: "node_modules"
      }
    })
  );
}
plugins.push(typescript({ tsconfig: "tsconfig.json" }));
plugins.push(
  // Transpile the code to our target - which is defined in .browserslistrc
  babel({
    exclude: "node_modules/**" // Don't transpile code other than our own
    // The rest of the config comes from babel.config.js
  })
);

const config = {
  external: (id, parentId, isResolved) => {
    return id.startsWith("core-js") || id.startsWith("regenerator-runtime");
  },
  input: "src/index.ts",
  output: {
    exports: "named", // We use named exports rather than "default"
    name: "sharedjs", // This is for the iife build
    sourcemap: true, // Generates a source map for the bundle.
    dir: "dist",
    format: format,
    entryFileNames: "index.[format].js"
  },
  plugins: plugins
};

export default config;
