// @maersk-global/shared-js/tests/lib/utilities.mock.js
export const decisions = {
  TABS: new Map([
    ["export", [44, 1, 3, 11, 15, 35, 25, 22]],
    ["import", [44, 4, 41, 39, 32, 5, 6, 16, 12, 31, 13]],
    ["pricing", [1, 44, 3, 40, 38, 4, 41, 39, 32, 5, 6, 22, 15, 11, 14, 33, 34, 16, 12, 31]],
    ["documentation", [1, 44, 3, 40, 38, 4, 41, 39, 32, 5, 6, 22, 15, 11, 25, 14, 33, 34, 16, 12, 31, 13]],
    ["log", [1, 44, 3, 40, 38, 4, 41, 39, 32, 5, 6, 22, 35, 15, 11, 25, 14, 33, 34, 16, 12, 31, 13]]
  ]),

  FEATURES: {
    EXPORT: new Map([
      ["cancel-booking", [1, 44, 38, 40, 3]],
      ["amend-booking", [1, 44, 38, 40, 3]],
      ["duplicate-booking", [1, 44, 38, 40, 3]],
      ["submit-si", [1, 3, 38, 40, 44, 15]],
      ["amend-draft", [15, 22]],
      ["submit-vgm", [1, 44, 38, 40, 3, 15]],
      ["view-draft-tpdoc", [44, 15, 22]],
      ["amend-final-tpdoc", [22]],
      ["issue-final-tpdoc", [22]],
      [
        "missing-invoice-details",
        [
          [1, 14],
          [1, 33],
          [1, 34],
          44,
          [38, 14],
          [38, 33],
          [38, 34],
          [40, 14],
          [40, 33],
          [40, 34],
          [3, 14],
          [3, 33],
          [3, 34],
          [15, 14],
          [15, 33],
          [15, 34],
          [11, 14],
          [11, 33],
          [11, 34],
          [22, 14],
          [22, 33],
          [22, 34]
        ]
      ],
      ["dg-document-submission", [1, 44, 38, 40, 3, 15]],
      ["submit-aes", [1, 44, 38, 40, 3, 15]],
      ["submit-customs-house-brokerage", [1, 44, 38, 40, 3]],
      ["add-vas", [1, 44, 38, 40, 3]],
      ["split-booking", [1, 44, 38, 40, 3]],
      ["add-new-export-party", [1, 44, 38, 40, 3]],
      ["bol-release", [22]],
      ["manage-references", [1, 44, 38, 40, 3, 15]]
    ]),

    IMPORT: new Map([
      ["surrender-b-l", [32, 39, 41, 4]],
      ["delivery-order-status", [32, 39, 41, 4, 31]],
      [
        "missing-invoice-details",
        [
          44,
          [32, 14],
          [32, 33],
          [32, 34],
          [39, 14],
          [39, 33],
          [39, 34],
          [41, 14],
          [41, 33],
          [41, 34],
          [4, 14],
          [4, 33],
          [4, 34],
          [5, 14],
          [5, 33],
          [5, 34],
          [6, 14],
          [6, 33],
          [6, 34],
          [31, 14],
          [31, 33],
          [31, 34],
          [16, 14],
          [16, 33],
          [16, 34],
          [12, 14],
          [12, 33],
          [12, 34],
          [22, 14],
          [22, 33],
          [22, 34]
        ]
      ],
      ["change-cy-to-sd", [44, 32, 39, 41, 4, 31]],
      ["add-customs-house-brokerage", [44, 32, 39, 41, 4]],
      ["add-new-import-party", [44, 32, 39, 41, 4]],
      ["bol-release", [32]],
      ["view-dnd", [4, 5, 6, 16, 31, 32, 39, 41, 44, 48, 49]]
    ]),

    PRICING: new Map([
      [
        "view-amount",
        [
          [1, 14],
          [1, 33],
          [1, 34],
          [3, 14],
          [3, 33],
          [3, 34],
          [4, 14],
          [4, 33],
          [4, 34],
          [5, 14],
          [5, 33],
          [5, 34],
          [6, 14],
          [6, 33],
          [6, 34],
          [11, 14],
          [11, 33],
          [11, 34],
          [12, 14],
          [12, 33],
          [12, 34],
          [15, 14],
          [15, 33],
          [15, 34],
          [16, 14],
          [16, 33],
          [16, 34],
          [22, 14],
          [22, 33],
          [22, 34],
          [31, 14],
          [31, 33],
          [31, 34],
          [32, 14],
          [32, 33],
          [32, 34],
          44
        ]
      ],
      [
        "edit-pricing",
        [3, [4, 14], [4, 34], [4, 34], 15, 22, 38, [39, 14], [39, 33], [39, 34], 40, [41, 14], [41, 33], [41, 34], 44]
      ]
    ]),
    SUBMIT_SI: new Map([
      ["submit-si", [1, 3, 38, 40, 44, 15]],
      ["resubmit-si", [1, 3, 38, 40, 44, 15]]
    ])
  }
};
